### ISSUE TITLE
Optionally support wheels

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-smithy#608]
It might be interesting for some subset of packages to be available as wheels. `conda-build` has support (guessing 3.x) for outputting wheels as noted in this [doc]( https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#output-type ). Doing this could help consolidate the wheel and conda-forge build projects a bit more, which would lend itself to a larger interested group working in the same ecosystem.

**Comment by isuruf** [comment]
For linux wheels, we'll have to use the manylinux1 docker container right?

**Comment by jakirkham** [comment]
It's certainly something we could do. Though I'd like to avoid using two docker images if possible.

**Comment by isuruf** [comment]
It's centos5 based while the conda-forge image is centos6 based.

**Comment by jakirkham** [comment]
Yep, I’m aware. There has been discussion about a manylinux2. Presumably it would be CentOS 6 based. We could start talking with people involved with manylinux1’s maintenance to see if we could agree on an image.

**Comment by jakirkham** [comment]
Should add that I'm mainly thinking these get published to [anaconda.org]( https://anaconda.org/ ) for now, but they could certainly be reused elsewhere. Possibly could publish them to other places if that is of interest too.

**Comment by jakirkham** [comment]
Would be good to hear some feedback from those close to the manylinux effort.

cc @ncoghlan @njsmith (others?)

**Comment by ncoghlan** [comment]
I'm a fan of evolving the `manylinux` baseline, but I'd like to switch it to a "target compatibility year" model on the next update.

That is, instead of it being `manylinux2`, a RHEL/CentOS 6 based version would instead be `manylinux2010`, then `manylinux2014` would be a RHEL/CentOS 7 base, etc.

I also think the compatibility detection section in https://www.python.org/dev/peps/pep-0513/#platform-detection-for-installers will require an update, such that we give installers a config file with the *compatibility profiles* in it. That is, the config file wouldn't say "This platform is manylinuxX compliant", but would instead say "For manylinuxX compliance, check for glibc Y.Z".

(Part of my inspiration for the latter idea is http://www.vfxplatform.com/ - once a means of checking a compatibility profile has been defined, the main thing new editions need to do is update the exact versions to be checked for)

**Comment by njsmith** [comment]
I think closer interactions between conda folks and wheel folks can only be good for both sides...

Regarding CentOS 6 and manylinux, everyone agrees that we need to update the setup and it's pretty straightforward to do, but someone needs to do the work and no-one's volunteered. There's a list of the required tasks here: https://mail.python.org/pipermail/distutils-sig/2017-April/030361.html

@ncoghlan Using calver here seems reasonable to me. I'm not sure I follow the part about the config file -- our current config file (the `import _manylinux` thing) is an escape hatch, mainly useful for distributors who want to be all grar go away I hate wheels. Otherwise, we currently define the glibc version → manylinux version table in the PEP text, which seems as useful as anything else, since every installer has to be updated when a new manylinux spec gets released anyway (we don't have any way to automagically distribute a config file to them).

**Comment by ncoghlan** [comment]
@njsmith I'm thinking something like certifi or pytz - a `manylinux` package on PyPI that gives you easy access to the kind of information the PEP describes. Even if installer devs write their own variant, they could use a common implementation in their test suite as a cross-check.

**Comment by jakirkham** [comment]
For some use cases, was recently asked about pip installing a package that I help maintain. The package was available as an sdist from PyPI. However the user ran into some glibc issues caused by build location vs. installing location discrepancy. In another use case, a user requested wheels for another project that I help maintain.

In pursuit of these use cases, generated PR ( https://github.com/conda-forge/numcodecs-feedstock/pull/17 ) and PR ( https://github.com/conda-forge/pyfftw-feedstock/pull/16 ). Both of these build the packages in question on CI and package them in both conda and wheel form thanks to some new features in conda-build. Tested building both of them locally and was able to get them to pip install into a minimal Python 3.6 environment and get them to work.

These particular cases work well IMHO as they don't really have any runtime dependencies that are other libraries. In the first case, everything is linked together into libraries that would be installed in site-packages regardless. In the second case, we already statically link a library to avoid some known issues encountered in the wild.

There's probably still some things to iron out like whether this is something others are comfortable pursuing in conda-forge, what other tooling changes need to be made (e.g. the upload script), how do we decide what packages should also build wheels, how do we handle more complex wheels (i.e. required shared libraries at runtime), how do we handle discrepancies in compatibility (e.g. CentOS 5 vs 6, macOS 10.6 vs 10.9), and probably others that I have thought of yet.

Anyways seems like a good time to get some other thoughts and feedback.

**Comment by jakirkham** [comment]
cc @conda-forge/core

**Comment by jakirkham** [comment]
Put together another PR ( https://github.com/conda-forge/python-snappy-feedstock/pull/12 ) to push on the dynamic library problem. However not totally sure how to proceed. Right now it assumes users have the dependencies on their system, which may not be reasonable. So would welcome feedback from wheel experts.

Guessing we need to use auditwheel (Linux), delocate (macOS), and ??? (Windows). Betting auditwheel runs into trouble as we are not using CentOS 5. Guessing delocate works, but we need to package it first. ( https://github.com/conda-forge/staged-recipes/issues/5316 ) Hoping that's not too difficult. Not sure what to do about Windows.

Also read the pynativelib proposal ( https://github.com/pypa/wheel-builders/pull/2 ), which is an interesting alternative. Though would need pointers on how that would work on different platforms and what sort of tooling needs to be employed.

cc @msarahan (as we're having a related conversation at `conda-build`)

**Comment by msarahan** [comment]
One thing that seems like an inevitable headache is that Anaconda has chosen to allow the newest language features (C++17 and such), and thus we need to ship the c++ runtime.  I think that's not something wheel builders want to do, but I might be wrong.  I think they'd rather use things like devtoolset so that they can keep older system libstdc++ and not ship it.  I don't know what impact that has on being able to use C++ 14/17/whatever - I know 11 works.

I need to spend some time with PEP517/518 to understand how conda-build fits in.  That seems key here, as it nicely defines the abstraction of a build system.  @njsmith what is the status of pynativelib?  That's also going to be very important to all of this, since it determines how much hacking/repacking of wheels goes on to fulfill native dependencies.

**Comment by jakirkham** [comment]
Guessing most people have seen this, but dropping it here for general reference. [PEP 571]( https://www.python.org/dev/peps/pep-0571/ ) is an in progress proposal for `manylinux2`, which would be a CentOS 6-based build image.

**Comment by jakirkham** [comment]
cc-ing @di, as we had talked about this briefly tonight. Would be interested in hearing your thoughts on this.
